woocommerce/woocommerce
# Admin list table [COT]

Adds a new list table implementation that works with COT, and otherwise strives to emulate the existing post-based admin list table.

![cot-admin-list-table](https://user-images.githubusercontent.com/3594411/167032599-8842b60f-adbe-480f-bb7b-04f0db286a73.png)

- Per the linked issue, we are not implementing all of the functionality in this PR (bulk actions, filters etc will not work).
- We're aiming to get close to the look and feel of the existing posts based table, but the result may not be identical and that's ok.
- In the above screenshot you may notice two **WooCommerce > Orders** menu entries: that's because, for testing purposes (remembering the new data-store is not 100% operational just yet):
    - I configured the WP Posts store as authoritative (therefore, the original admin list table and its menu entry remain accessible).
    - I temporarily tweaked [this line of code](https://github.com/woocommerce/woocommerce/pull/32864/files#diff-1c0f35a6fb307b4248784e1f47ff1e71754b02fd1315315e44b5b5eb725f30b4R319) by adding `true ||` to the start of the condition, to make the new admin list table render anyway.
    - With a few exceptions, the new implementation mostly doesn't care which data store is authoritative (both implement the same interface), which is why this approach essentially works.

Closes #32385.

### Testing

1. For completeness, you may wish to follow the steps in [the linked issue](https://github.com/woocommerce/woocommerce/issues/32385) to enable and hydrate the new order table (though this probably isn't strictly required, since at time of writing we can't read from the new store in any case—however, that will change in the future).
2. Per the above notes, temporarily tweak the code (inside the `WC_Admin_Menus::orders_menu()` method) as suggested and set the WP Posts store as default via **WooCommerce > Settings > Advanced > Custom Data Stores**.
3. Navigate to the **new** admin list table (look for `admin.php?page=wc-orders` in the URL).
4. Compare and contrast with the original post-based admin list table, remembering per the above notes that not everything is fully functional just yet.
5. Check the no-results page. There are two ways to see this:
    - At time of writing, you can do this by setting COT to authoritative and removing the temporary shim from `WC_Admin_Menus::orders_menu()`.
    - Or, keeping the shim in place (and keeping the post table as authoritative) you can instead simply hide your post table orders with a query as follows (to undo, just reverse the post type values):

```sh
wp db query "UPDATE $(wp db prefix)posts SET post_type = 'shop_order_hidden' WHERE post_type = 'shop_order'" ;\
wp cache flush
```

### Notes/questions

- You may need to rebuild assets and hard-refresh your browser for the relevant CSS+JS changes to show up.
- We may (or may not) wish to expand e2e-core-tests to cover this area (I don't believe the existing list table is covered, either); but it probably makes sense to defer that until we're at a more advanced stage of the COT project.
- Are we happy with the placement inside the `Admin\Internal` namespace (sidenote: its location is the reason an automated action applies the `focus: react admin` label) or might we prefer to place it in some other namespace?
- Admin UI for the order editor is not yet implemented (so following the link to add a new order will fail when COT is authoritative).
- Normally, if some orders are pending, a number will display next to the Orders menu entry, as in the following screenshot (something similar happens with the new WC Admin navigation feature) ... this will not currently work when COT is authoritative, because the necessary logic to perform order counts is not yet implemented.

![orders-posts-needing-attention-bubble](https://user-images.githubusercontent.com/3594411/167725095-898be602-b72a-43d3-9915-af479dedf4c5.png)
